### PR TITLE
provides Loadbalancer improvement and reimplementation

### PR DESCRIPTION
**Pull request by OlegDokuka**
This PR provides a rework of existing Loadbalance strategies and uses the built-in into RSocket-Core

### CODE REVIEW

File: rsocket-routing-broker-spring/src/main/java/io/rsocket/routing/broker/spring/BrokerAutoConfiguration.java [hunk 1]
```diff
@@ -137,32 +142,27 @@ public ReactorResourceFactory reactorResourceFactory() {
...
-		return new WeightedRSocketFactory();
-	}
-
 	// TODO: pick LoadBalancer algorithm via tags
 	@Bean
 	@ConditionalOnMissingBean
 	@ConditionalOnProperty(prefix = BROKER_PREFIX, name = "default-load-balancer", havingValue = "weighted")
-	public WeightedLoadBalancer.Factory weightedLoadBalancerFactory() {
-		return new WeightedLoadBalancer.Factory();
+	public WeightedLoadbalanceStrategy weightedLoadBalancerFactory() {
```
Comment (spencergibb) [inline rsocket-routing-broker-spring/src/main/java/io/rsocket/routing/broker/spring/BrokerAutoConfiguration.java:149]: nit: bean name to match

File: rsocket-routing-broker-spring/src/main/java/io/rsocket/routing/broker/spring/BrokerAutoConfiguration.java [hunk 2]
```diff
@@ -137,32 +142,27 @@ public ReactorResourceFactory reactorResourceFactory() {
...
+		return new WeightedLoadbalanceStrategy();
 	}
 
 	@Bean
 	@ConditionalOnMissingBean
 	@ConditionalOnProperty(prefix = BROKER_PREFIX, name = "default-load-balancer", havingValue = "roundrobbin", matchIfMissing = true)
-	public RoundRobinLoadBalancer.Factory roundRobinLoadBalancerFactory() {
-		return new RoundRobinLoadBalancer.Factory();
+	public RoundRobinLoadbalanceStrategy roundRobinLoadBalancerFactory() {
```
Comment (spencergibb) [inline rsocket-routing-broker-spring/src/main/java/io/rsocket/routing/broker/spring/BrokerAutoConfiguration.java:156]: nit: bean name to match

File: rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/FluxStats.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+class FluxStats extends Flux<Payload> implements Subscription, Subscriber<Payload> {
+
+	final Flux<Payload> upstream;
+	final Stats stats;
+
+	Subscription                    s;
+	CoreSubscriber<? super Payload> actual;
+
+	volatile int                                      state;
+	static final AtomicIntegerFieldUpdater<FluxStats> STATE =
```
Comment (spencergibb) [inline rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/FluxStats.java:36]: I'm not familiar with `AtomicIntegerFieldUpdater`. Why choose it over `AtomicBoolean`?

Comment (spencergibb) [reply rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/FluxStats.java:36]: 3 states

File: rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/ResolvingRSocket.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import io.rsocket.frame.FrameType;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Operators;
+import reactor.util.context.Context;
+
+public final class ResolvingRSocket extends ResolvingOperator<RSocket>
```
Comment (spencergibb) [inline rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/ResolvingRSocket.java:32]: Seems weird that we have to extend a package private class. If it's meant to be used outside, maybe it should be public

File: rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/ResolvingRSocket.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    implements CoreSubscriber<RSocket>, RSocket {
+
+  final Publisher<RSocket> rSocketMono;
+
+  volatile Subscription s;
+
+  static final AtomicReferenceFieldUpdater<ResolvingRSocket, Subscription> S =
```
Comment (spencergibb) [inline rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/ResolvingRSocket.java:39]: Again, why not just an `AtomicReference`?

File: rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/FluxStats.java (hunk 3 above)
Comment (OlegDokuka) [reply rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/FluxStats.java:36]: Basically, to reduce the number of allocation, there is no AtomicBooleanFieldUpdater or soon thus to keep the number of the allocated objects as small as possible, uses that one

File: rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/ResolvingRSocket.java (hunk 5 above)
Comment (OlegDokuka) [reply rsocket-routing-broker/src/main/java/io/rsocket/loadbalance/ResolvingRSocket.java:39]: > Again, why not just an AtomicReference?

The same as in the previous comments. `AtomicReference` - is an object, every object puts more load on the GC, the less of them we allocate the higher performance we get.

**Review by rstoyanchev — COMMENTED** [review]
There is a package cycle between the `locator` and `rsocket` packages with `locator` depending on `MulticastRSocket` and `rsocket` depending on `RSocketLocator`. Consolidating these into one package would probably make sense. The seem closely related.

**Comment by spencergibb** [comment]
I'll fix the tangle and we can make changes if rsocket java we can adapt here.

**Pull request merged**